File: cmd/mattermost/commands/permissions_test.go
// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

package commands

import (
	"testing"

	"github.com/mattermost/mattermost-server/v5/shared/i18n"
	"github.com/stretchr/testify/assert"
)

func TestPermissionsExport_rejectsUnlicensed(t *testing.T) {
	th := Setup(t).InitBasic()
	defer th.TearDown()

	actual, _ := th.RunCommandWithOutput(t, "permissions", "export")
	assert.Contains(t, actual, i18n.T("cli.license.critical"))
}

func TestPermissionsImport_rejectsUnlicensed(t *testing.T) {
	th := Setup(t).InitBasic()
	defer th.TearDown()

	actual, _ := th.RunCommandWithOutput(t, "permissions", "import")

	assert.Contains(t, actual, i18n.T("cli.license.critical"))
}

File: webapp/channels/src/packages/mattermost-redux/src/constants/permissions.test.ts
// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import Permissions, {CHANNEL_WRITE_PERMISSIONS} from 'mattermost-redux/constants/permissions';

describe('CHANNEL_WRITE_PERMISSIONS', () => {
    it('mirrors isChannelWritePermission in server/channels/app/channel_access.go', () => {
        expect(CHANNEL_WRITE_PERMISSIONS.size).toBe(38);
    });

    it('contains only known permissions', () => {
        const known = new Set(Object.values(Permissions).filter((value): value is string => typeof value === 'string'));

        for (const permission of CHANNEL_WRITE_PERMISSIONS) {
            expect(known).toContain(permission);
        }
    });

    it('excludes read access and policy administration', () => {
        expect(CHANNEL_WRITE_PERMISSIONS.has(Permissions.READ_CHANNEL)).toBe(false);
        expect(CHANNEL_WRITE_PERMISSIONS.has(Permissions.READ_CHANNEL_CONTENT)).toBe(false);
        expect(CHANNEL_WRITE_PERMISSIONS.has(Permissions.MANAGE_CHANNEL_ACCESS_RULES)).toBe(false);
    });
});
